packages/core/src/browser/shell/tab-bar-toolbar.tsx

```diff
@@ -272,11 +295,13 @@ export class TabBarToolbarRegistry implements FrontendApplicationContribution {
      *
      * By default returns with all items where the command is enabled and `item.isVisible` is `true`.
      */
-    visibleItems(widget: Widget): TabBarToolbarItem[] {
+    visibleItems(widget: Widget): Array<TabBarToolbarItem | ReactTabBarToolbarItem> {
         const result = [];
         for (const item of this.items.values()) {
-            if (this.commandRegistry.isVisible(item.command, widget) &&
-                (!item.when || this.contextKeyService.match(item.when, widget.node))) {
+            const visible = TabBarToolbarItem.is(item)
+                ? this.commandRegistry.isVisible(item.command, widget)
+                : (!item.isVisible || item.isVisible(widget));
```

Comment:
`: item.isVisible && item.isVisible(widget);` ?

Reply:
I think the current change is correct. You either have the `isVisible` defined, then it has to be evaluated to `true`, or you do not have this constraint; hence your item will always be visible.

Reply:
Ah ok, I made it default to false on my side, read that a bit quickly, you are right :)